File: golang/solutions/z2.go
package solutions

import (
	"aoc/utils"
	"fmt"
	"math"
	"strconv"
	"strings"
)

type gameData struct {
	red   int
	green int
	blue  int
}

type gameInfo struct {
	id    int
	games []gameData
}

func parseGameInfo(line string) gameInfo {
	arr := strings.Split(line, ":")
	head := strings.Split(arr[0], " ")
	id, _ := strconv.Atoi(head[1])

	games := make([]gameData, 0)

	for _, game := range strings.Split(arr[1], ";") {
		colors := strings.Split(game, ",")
		gd := gameData{}
		for _, color := range colors {
			color = strings.Trim(color, " ")
			data := strings.Split(color, " ")
			switch data[1] {
			case "red":
				gd.red, _ = strconv.Atoi(data[0])
				continue
			case "blue":
				gd.blue, _ = strconv.Atoi(data[0])
				continue
			case "green":
				gd.green, _ = strconv.Atoi(data[0])
				continue
			}
		}
		games = append(games, gd)
	}

	return gameInfo{
		id:    id,
		games: games,
	}
}

func isValidGame(data []gameData) bool {
	limit := gameData{red: 12, green: 13, blue: 14}
	for _, game := range data {
		if game.blue > limit.blue || game.red > limit.red || game.green > limit.green {
			return false
		}
	}
	return true
}

func getMaxValue(data []gameData) gameData {
	result := gameData{}
	for _, game := range data {
		result.red = int(math.Max(float64(game.red), float64(result.red)))
		result.green = int(math.Max(float64(game.green), float64(result.green)))
		result.blue = int(math.Max(float64(game.blue), float64(result.blue)))
	}
	return result
}

func Solve2() {
	data := utils.ReadLines("../inputs/z2.txt")
	sum := 0
	pow := 0
	for _, line := range data {
		game := parseGameInfo(line)

		if isValidGame(game.games) {
			sum = game.id + sum
		}

		max := getMaxValue(game.games)
		pow += max.blue * max.green * max.red
	}

	fmt.Printf("Result#1 %d\n", sum)
	fmt.Printf("Result#2 %d\n", pow)
}

File: ts/z2.ts
import fs from 'fs';

async function parseLinesFromFile(filePath: string) {
    const content = fs.readFileSync(filePath).toString('utf-8')
    return content.split('\n').filter(notEmpty => notEmpty);
}

type GameData = {
    red: number;
    green: number;
    blue: number;
}

type GameInfo = {
    id: number;
    games: GameData[];
};

function parseData(data: string): GameData {
   const gd = data.split(',').map(x => x.trim())
   .map(x => x.split(' '))
   .reduce((acc, [num, str]) => {
    (acc as any)[str] = +num;
    return acc;
   }, {
    red: 0,
    green: 0,
    blue: 0,
   });

   return gd;
}

function getGame(str: string): GameInfo {
    const [game, rawGames] = str.split(':');
    const id = +game.split(' ').pop()!;
    const games = rawGames.split(';').map(parseData);

    return {
        id,
        games,
    };
}

function isPossible(games: GameData[], limit: GameData = {red: 12, green: 13, blue: 14}) {
    return games.every(game => game.red <= limit.red && game.blue <= limit.blue && game.green <= limit.green);
}

function solvePart1(all: GameInfo[]) {
    const valid = all.filter(gd => isPossible(gd.games));
    return valid.reduce((acc, info) => acc + info.id, 0);
}

function solvePart2(all: GameInfo[]) {
    const max = all.map(x => x.games.reduce((max, gd) => {
      max.blue = Math.max(gd.blue, max.blue);
      max.green = Math.max(gd.green, max.green);
      max.red = Math.max(gd.red, max.red);
      
      return max;
    }));

    return max.reduce((pow, gd) => {
        pow += gd.blue * gd.red * gd.green;
        return pow;
    }, 0);
}

async function main() {
    const data = await parseLinesFromFile('../inputs/z2.txt');
    const all = data.map(getGame);
   

    console.log(`result#1 ${solvePart1(all)}`);
    console.log(`result#2 ${solvePart2(all)}`);
}

main();